src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -61,19 +79,37 @@ public static class ActivatorUtilities
                     {
                         bestLength = length;
                         bestMatcher = matcher;
+                        multipleBestLengthFound = false;
+                    }
+                    else if (bestLength == length)
+                    {
+                        multipleBestLengthFound = true;
                     }
 
                     seenPreferred |= isPreferred;
                 }
+
+                if (bestLength != -1)
+                {
+                    if (multipleBestLengthFound)
+                    {
+                        throw new InvalidOperationException(SR.Format(SR.MultipleCtorsFoundWithBestLength, instanceType, bestLength));
+                    }
+
+                    return bestMatcher.CreateInstance(provider);
+                }
             }
 
-            if (bestLength == -1)
+            var argumentTypes = new Type[parameters.Length];
+            for (int i = 0; i < argumentTypes.Length; i++)
             {
-                string? message = $"A suitable constructor for type '{instanceType}' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.";
-                throw new InvalidOperationException(message);
+                argumentTypes[i] = parameters[i].GetType();
```

Comment:
Can one of the `parameters` be `null`? If so, this will NRE.

Reply:
no parameters cannot be null on the CreateInstance signature:

https://github.com/dotnet/runtime/blob/2b07c8c46523b816b1eac5c8af092e52b20a69d2/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L29-L32

Reply:
We are checking for `null` elsewhere:

https://github.com/dotnet/runtime/blob/3435b39194137676d9b2da12c8ebd21d69cd06b9/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L345

Nothing is guaranteeing the caller doesn't pass `null` inside this array. The signature just tells the compiler to log a warning if they do. The caller can suppress that using `!`, or they could be using `netstandard2.0` which won't give them nullability warnings.

We shouldn't be throwing NullReferenceException here.